Bugfix/ Ensure that the stored file is the same as the one passed from client

maxr96 opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] ~Client: I added multiple integration tests (Jest) related to the features~
- [x] ~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~
- [x] ~Client: I documented the TypeScript code using JSDoc style.~
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] ~Client: I translated all the newly inserted strings into German and English~

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
We don't want user to loose their files and we notify them if the file was not stored correctly due to some internal error on the server.
Fixes #1097

### Description
<!-- Describe your changes in detail -->
We now check if the hash sum of the file passed from the client is the same as hash sum of the file stored on the server. Additionally, we also check if the file content is not empty to avoid storing empty files on the server. 
If the submission of the file upload exercise failed error message is shown.
![Screenshot_2019-12-04 File Upload Exercises(1)](https://user-images.githubusercontent.com/17262376/70186156-8f912080-16eb-11ea-985f-026b6c2e6c41.png)

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Overview
3. Select Course with active File Upload Exercise
4. Start File Upload Exercise
5. Submit a file (no errors shown)

maxr96 commented:
> File upload still works and empty files are being rejected.
> 
> The exception message is not passed on to the frontend, right? I'm only getting the generic `Your submission for file test.png failed!` alert in the frontend when submitting an empty file.

Thanks for the review @HanyaElhashemy and @madwau. Yes the error is still generic. In my opinion the user should not see specific details about the error.

madwau commented:
Hm, if the uploaded file is empty, why not let the user know about it so they know it's their fault and they can fix it? Off the top of my head I don't see any security implications here.

lovaiible commented on src/main/java/de/tum/in/www1/artemis/service/FileUploadSubmissionService.java:
```diff
@@ -184,11 +186,21 @@ public FileUploadSubmission save(FileUploadSubmission fileUploadSubmission, Mult
+        if (!multipartFileHash.equals(storedFileHash)) {
+            throw new IOException("The file " + file.getName() + "could not be stored");
+        }
+
```
Since I could not highlight the Transactional annotation above, I'm writing my comment here :) 
Do we need `@Transactional(rollbackFor = Exception.class)` for the save method? If not, then please remove it.

maxr96 replied:
This change is a part of a PR https://github.com/ls1intum/Artemis/pull/1107, so I didn't do it here yet. But now it is done

maxr96 commented:
> Hm, if the uploaded file is empty, why not let the user know about it so they know it's their fault and they can fix it? Off the top of my head I don't see any security implications here.

Thank you for the suggestion @madwau . I first thought that it is not possible to upload empty file on purpose, but I was able to locally create empty file and receive the error. I created a new `EmptyFileException` for this case and now the error should be correctly propagated to the client.

maxr96 commented on src/main/java/de/tum/in/www1/artemis/service/FileUploadSubmissionService.java:
```diff
@@ -184,11 +186,21 @@ public FileUploadSubmission save(FileUploadSubmission fileUploadSubmission, Mult
+        }
+        var multipartFileHash = DigestUtils.md5Hex(file.getInputStream());
         // check if we already had file associated with this submission
         fileUploadSubmission.onDelete();
```
Yes this will leave the user with no files, which we should not do. I tried to simply move it, but the problem is that `onDelete` will remove the latest file stored in the folder of this submission and in this case it is newly uploaded file. I will check how can I do this in a different way.

maxr96 replied:
I was able to fix this. Now we don't delete the directory were files are stored, but only the old file itself. So as you suggested the old file is deleted only after we have checked the new file is stored properly.

beatrisbu commented:
> Nice fix, especially saving the last successful file after an unsuccessful file upload. However, could you also show it on the GUI? It still says "you have no submission". I have to navigate back to the overview and click the "view submission" button again in order to see it after an unsuccessful submit. It would be good if the "you have submitted a file xmin ago" description does not disappear after an unsuccessful submit.
> 
> Screenshot 1: The view after an unsuccessful file upload. Note that the file successfully uploaded before disappears from the GUI
> ![image](https://user-images.githubusercontent.com/50635005/70380340-cbb9c080-1939-11ea-8bb5-48767856def9.png)
> Screenshot 2: Says no file submitted although before the unsuccessful submission there was a successful submission. You have to click view submission again to see it
> ![image](https://user-images.githubusercontent.com/50635005/70380351-1dfae180-193a-11ea-89d0-e1716f372c18.png)

I agree with the first point. 

I have a question about the second one - Where is the user supposed to see that a file is submitted on the second screenshot? We have the same behavior with Text submissions as well, as far as I can see. The "results" get updated after a grading has been received, right?

beatrisbu commented:
Nice work! The functionality is there. I can successfully upload a file and get a message validating that. When I submit an empty file, I get a notification about the problem as well. 

I agree with Hanya that it would be nice if there is an option to not overwrite the successful upload on the "View submission page" when uploading an empty file.

HanyaElhashemy commented:
> > Nice fix, especially saving the last successful file after an unsuccessful file upload. However, could you also show it on the GUI? It still says "you have no submission". I have to navigate back to the overview and click the "view submission" button again in order to see it after an unsuccessful submit. It would be good if the "you have submitted a file xmin ago" description does not disappear after an unsuccessful submit.
> > Screenshot 1: The view after an unsuccessful file upload. Note that the file successfully uploaded before disappears from the GUI
> > ![image](https://user-images.githubusercontent.com/50635005/70380340-cbb9c080-1939-11ea-8bb5-48767856def9.png)
> > Screenshot 2: Says no file submitted although before the unsuccessful submission there was a successful submission. You have to click view submission again to see it
> > ![image](https://user-images.githubusercontent.com/50635005/70380351-1dfae180-193a-11ea-89d0-e1716f372c18.png)
> 
> I agree with the first point.
> 
> I have a question about the second one - Where is the user supposed to see that a file is submitted on the second screenshot? We have the same behavior with Text submissions as well, as far as I can see. The "results" get updated after a grading has been received, right?

Yes, you are right. Thanks for drawing my attention. I updated my comment!
It is only missing on the view submission page, as seen in screenshot 1 :)

maxr96 commented:
> Nice fix, especially saving the last successful file after an unsuccessful file upload. However, could you also show it on the GUI? I have to navigate back to the overview and click the "view submission" button again in order to see it after an unsuccessful submit. It would be good if the "you have submitted a file xmin ago" description does not disappear after an unsuccessful submit.
> 
> Screenshot 1: The view after an unsuccessful file upload. Note that the file successfully uploaded before disappears from the GUI
> ![image](https://user-images.githubusercontent.com/50635005/70380340-cbb9c080-1939-11ea-8bb5-48767856def9.png)

Thank you for reviewing my PR @beatrisbu and @HanyaElhashemy. Good point about still showing previous submission on incorrect submission. Now it is fixed.
![Screenshot_2019-12-08 File Upload Exercises](https://user-images.githubusercontent.com/17262376/70392496-fb6fd380-19e0-11ea-8834-3a04115ec03f.png)

HanyaElhashemy commented:
Good Fix! Thanks for responding to my comment ^^

ungaralex commented on src/main/webapp/i18n/de/error.json:
```diff
@@ -20,6 +20,8 @@
-        "groupNotFound": "Die von dir eingegebene Gruppe wurde nicht gefunden."
+        "groupNotFound": "Die von dir eingegebene Gruppe wurde nicht gefunden.",
+        "cantSaveFile": "Die hochgeladene Datei {{ fileName }} konnte nicht gespeichert werden.",
+        "emptyFile": "Die hochgeladene Datei {{ fileName }} ist leer. Bitte laden Sie eine andere Datei hoch"
```
We should be consistent with the salutation. Either use "du" or "Sie" in German. I think we are already using "du" most of the times, so I would change it here.

maxr96 replied:
You are right here. I use `du` now

ungaralex commented on src/test/java/de/tum/in/www1/artemis/FileUploadSubmissionIntegrationTest.java:
```diff
@@ -109,6 +109,15 @@ public void submitFileUploadSubmission() throws Exception {
 
+    @Test
+    @WithMockUser(value = "student3")
+    public void submitFileUploadSubmission_emptyFileContent() throws Exception {
```
Just a couple of ideas for additional tests:
Do we also have a test checking if the file is actually stored in the filesystem? Especially with the new hash comparison, we could additionally check if the stored file has the same hash. 
Also, another idea: Mock the saveFileForSubmission method and store the wrong file forcing the IOException caused by different hashes.

maxr96 replied:
I now check that the file is actually stored in the `submitFileUploadSubmission` test case.
I also thought about mocking the `saveFileForSubmission` method, but I didn't find any way to mock private method and I didn't want to make it public only for testing purposes.

Minh13 commented:
Hi, I tested it and it works as intedended for files/ empty files but I obviously wasn't able to reproduce an internal server error. Also well done keeping the PR manageable !

maxr96 commented:
Thank you for the reviews @ungaralex and @Minh13 . I implemented the suggested changes for translations and extended the file upload submission test to also check if the stored file is not empty

maxr96 commented:
> Thank you for addressing the issues I mentioned @maxr96 :)
> 
> Just a few things that I noticed while testing the new changes:
> 
>     * The submit button is looks like it's disabled after submitting a file (even if it's not)
>       ![image](https://user-images.githubusercontent.com/33764166/70460207-12ccc080-1ab6-11ea-8572-d0d9e437474c.png)
> 
>     * I could submit an empty PDF file, the file does not overwrite my old submission, but the application shows that I have successfully uploaded the file
>       ![image](https://user-images.githubusercontent.com/33764166/70460326-51627b00-1ab6-11ea-9149-b06b28941cc2.png)
> 
> 
> Empty file:
> ![image](https://user-images.githubusercontent.com/33764166/70460337-56bfc580-1ab6-11ea-9194-582fc0fc19f2.png)



> Thank you for addressing the issues I mentioned @maxr96 :)
> 
> Just a few things that I noticed while testing the new changes:
> 
>     * The submit button is looks like it's disabled after submitting a file (even if it's not)
>       ![image](https://user-images.githubusercontent.com/33764166/70460207-12ccc080-1ab6-11ea-8572-d0d9e437474c.png)
> 
>     * I could submit an empty PDF file, the file does not overwrite my old submission, but the application shows that I have successfully uploaded the file
>       ![image](https://user-images.githubusercontent.com/33764166/70460326-51627b00-1ab6-11ea-9149-b06b28941cc2.png)
> 
> 
> Empty file:
> ![image](https://user-images.githubusercontent.com/33764166/70460337-56bfc580-1ab6-11ea-9194-582fc0fc19f2.png)

Thank you for the review @lovaiible. 
- The submit button is indeed disabled until the file input is empty. It enables when you select a new file yo upload
- I tried to do the same with empty `pdf` file and I saw the error. Can you maybe give more details on how this happened? Did you check the network panel in the Developer Tools? You should see a 400 response from the server
![Screenshot_2019-12-09 File Upload Exercises](https://user-images.githubusercontent.com/17262376/70461127-3bee5080-1ab8-11ea-8ce6-fdaef04a1f0d.png)
![Screenshot_20191209_191616](https://user-images.githubusercontent.com/17262376/70461279-9b4c6080-1ab8-11ea-8766-e37310ea1b38.png)

krusche commented on src/main/java/de/tum/in/www1/artemis/service/FileService.java:
```diff
@@ -163,7 +163,7 @@ public String actualPathForPublicPath(String publicPath) {
             final var exerciseId = Long.parseLong(shouldBeExerciseId);
             final var submissionId = Long.parseLong(shouldBeSubmissionId);
-            return FileUploadSubmission.buildFilePath(exerciseId, submissionId);
+            return FileUploadSubmission.buildFilePath(exerciseId, submissionId) + filename;
```
how can this work for old files?

maxr96 replied:
For file upload submission we only use this method when deleting file associated with the submission. Before it was removing this file with the directory it was located in, now it will remove only the file itself. So this will not affect old files

ungaralex commented:
I have an additional idea: We could also store the file hash in the database, so that even in the worst case, students could prove that they submitted a specific file by comparing the hashes. What do you think @maxr96 @krusche ?

maxr96 commented:
> I have an additional idea: We could also store the file hash in the database, so that even in the worst case, students could prove that they submitted a specific file by comparing the hashes. What do you think @maxr96 @krusche ?

That is actually a good idea. We can definitely store the hashes, but I am not sure how the tutors/instructors can see them afterwards. So in this case student will have to send the submitted file manually to the tutor and they will do the hash check, right? 
I think now the worst case is the file being accidentally deleted, but at least the file path on the submission is stored. Tutor/instructor will still be able to see the fact of submission, only the file link will not work. In this case storing the hash will only help to ensure that no content was changed since the last submission. But I guess it could also happen that the student still does some changes and then he can't prove his submission anymore

jpbernius commented on src/main/webapp/i18n/de/error.json:
```diff
@@ -20,6 +20,8 @@
-        "groupNotFound": "Die von dir eingegebene Gruppe wurde nicht gefunden."
+        "groupNotFound": "Die von dir eingegebene Gruppe wurde nicht gefunden.",
+        "cantSaveFile": "Die hochgeladene Datei {{ fileName }} konnte nicht gespeichert werden.",
+        "emptyFile": "Die hochgeladene Datei {{ fileName }} ist leer. Bitte lädst du eine andere Datei hoch"
```
```suggestion
        "emptyFile": "Die hochgeladene Datei {{ fileName }} ist leer. Bitte überprüfe die Datei und versuche es erneut."
```

jpbernius commented on src/main/webapp/i18n/en/error.json:
```diff
@@ -18,8 +18,10 @@
-        "groupNotFound": "The group you entered was not found."
+        "groupNotFound": "The group you entered was not found.",
+        "cantSaveFile": "The uploaded file {{ fileName }} could not be saved.",
+        "emptyFile": "The uploaded file {{ fileName }} is empty. Please upload another file"
```
```suggestion
        "emptyFile": "The uploaded file {{ fileName }} is empty. Please check the file and try again."
```
